File: framework/languages/dotnet/e2e/RegistrationCodec/Server/Main/DispatchFilters.cs
using RegistrationCodec.Server.Main.Infrastructure;
using Zlink.Framework.Contracts.Handlers;

namespace RegistrationCodec.Server.Main;

internal sealed class FirstFilter(EvidenceStore evidence) : IZLinkHandlerFilter
{
    public async ValueTask<object?> InvokeAsync(
        ZLinkHandlerInvocation invocation,
        ZLinkHandlerDelegate next,
        CancellationToken cancellationToken)
    {
        evidence.Add($"filter|name=first|phase=before|packet={invocation.PacketName}");
        var result = await next(cancellationToken);
        evidence.Add($"filter|name=first|phase=after|packet={invocation.PacketName}");
        return result;
    }
}

internal sealed class SecondFilter(EvidenceStore evidence) : IZLinkHandlerFilter
{
    public async ValueTask<object?> InvokeAsync(
        ZLinkHandlerInvocation invocation,
        ZLinkHandlerDelegate next,
        CancellationToken cancellationToken)
    {
        evidence.Add($"filter|name=second|phase=before|packet={invocation.PacketName}");
        var result = await next(cancellationToken);
        evidence.Add($"filter|name=second|phase=after|packet={invocation.PacketName}");
        return result;
    }
}

File: framework/languages/node/e2e/RegistrationCodec/Server/Main/Handlers/dispatch-filters.ts
import { Injectable } from '@nestjs/common';
import type { ZLinkHandlerDelegate, ZLinkHandlerFilter, ZLinkMessageContext } from '@zlink-systems/framework';
import { EvidenceStore } from '../Infrastructure/evidence-store';

@Injectable()
export class FirstFilter implements ZLinkHandlerFilter {
  constructor(private readonly evidence: EvidenceStore) {}

  async invoke(context: ZLinkMessageContext, next: ZLinkHandlerDelegate): Promise<unknown> {
    this.evidence.add(`filter|name=first|phase=before|packet=${context.packetName ?? '<null>'}`);
    const result = await next();
    this.evidence.add(`filter|name=first|phase=after|packet=${context.packetName ?? '<null>'}`);
    return result;
  }
}

@Injectable()
export class SecondFilter implements ZLinkHandlerFilter {
  constructor(private readonly evidence: EvidenceStore) {}

  async invoke(context: ZLinkMessageContext, next: ZLinkHandlerDelegate): Promise<unknown> {
    this.evidence.add(`filter|name=second|phase=before|packet=${context.packetName ?? '<null>'}`);
    const result = await next();
    this.evidence.add(`filter|name=second|phase=after|packet=${context.packetName ?? '<null>'}`);
    return result;
  }
}
